library/core/src/array/mod.rs

```diff
@@ -393,7 +393,68 @@ impl<T, const N: usize> [T; N] {
     where
         F: FnMut(T) -> U,
     {
-        use crate::mem::MaybeUninit;
+        use crate::mem::{align_of, forget, size_of, transmute_copy, ManuallyDrop, MaybeUninit};
+
+        if align_of::<T>() == align_of::<U>() && size_of::<T>() == size_of::<U>() {
+            // this branch allows reuse of the original array as a backing store
```

Comment:
Suggestion: this is the kind of thing that tends to be somewhat fragile.  It would be good to have an LLVM codegen test that optimizations actually do what this is trying to do.  (And if LLVM can't currently do it, I'm not sure it's worth having extra code.)

Reply:
I've attached a test which should should change which code paths is run based on this branch, but I'm not sure how to properly annotate it.